AX: Interop 2024 WPT failure "fallback role w/ region with no label"

fleizach opened the pull request:
#### 5899e80acec9d3dee7328c06a23ab12da5124310
<pre>
AX: Interop 2024 WPT failure &quot;fallback role w/ region with no label&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=282271">https://bugs.webkit.org/show_bug.cgi?id=282271</a>
&lt;<a href="https://rdar.apple.com/138860789">rdar://138860789</a>&gt;

Reviewed by Tyler Wilcock.

aria-roles can have multiple tokens, but not every token should be respected.
Provide the ability to skip a role if the right condition is met.

Canonical link: <a href="https://commits.webkit.org/289531@main">https://commits.webkit.org/289531@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8350a21da1c9bd5d26c7dde27eaaf4e75e1045d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87209 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6719 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41569 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92070 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37949 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89259 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14788 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67407 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25140 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90211 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5362 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78935 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47727 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5135 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33314 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37065 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75618 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34190 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93955 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14371 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10466 "") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76208 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14575 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74791 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75411 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18556 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19756 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18192 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7300 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14390 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19683 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14135 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17578 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15916 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2810,9 +2810,18 @@ AccessibilityRole AccessibilityNodeObject::determineAriaRoleAttribute() const
+        if (nextRole != role)
+            role = nextRole;
+        else
+            role = AccessibilityRole::Unknown;
```
The reason I stayed away from this one was because technically we'd need to loop here, as `nextRole` may be a different role which also has constraints (e.g. maybe the next one is role `presentation`, which we need to not respect if the element is focusable, but we wouldn't do that check as things are now — or maybe we need to `remapAriaRoleDueToParent` on `nextRole`, etc).

That being said, I think the scenarios where this matters are gonna be super rare, and this overall is an improvement, so I think we can save these other things for a later PR.

Just a couple minor things — I think there's an extra newline after `if ((role == AccessibilityRole::LandmarkRegion || role == AccessibilityRole::Form) && !hasAccNameAttribute()) {`, and second I wonder if the comment could be on one line since it will still be shorter than the line of code that comes after it:

```cpp
// Fallback-roles - If a region has no label, but does have a fallback role, use that instead.
auto nextRole = ariaRoleToWebCoreRole(ariaRole, [] (const AccessibilityRole& skipRole) {
```

fleizach committed:
AX: Interop 2024 WPT failure "fallback role w/ region with no label"
https://bugs.webkit.org/show_bug.cgi?id=282271
<rdar://138860789>

Reviewed by Tyler Wilcock.

aria-roles can have multiple tokens, but not every token should be respected.
Provide the ability to skip a role if the right condition is met.

Canonical link: https://commits.webkit.org/289531@main
